tooling/nargo_cli/src/cli/debug_cmd.rs

```diff
@@ -173,66 +250,168 @@ fn run_async(
     prover_name: &str,
     witness_name: &Option<String>,
     target_dir: &PathBuf,
-) -> Result<(), CliError> {
+) -> Result<DebugResult, CliError> {
     use tokio::runtime::Builder;
     let runtime = Builder::new_current_thread().enable_all().build().unwrap();
 
     runtime.block_on(async {
         println!("[{}] Starting debugger", package.name);
-        let (return_value, witness_stack) =
-            debug_program_and_decode(program, package, prover_name)?;
-
-        if let Some(solved_witness_stack) = witness_stack {
-            println!("[{}] Circuit witness successfully solved", package.name);
-
-            if let Some(return_value) = return_value {
-                println!("[{}] Circuit output: {return_value:?}", package.name);
+        let debug_result = debug_program_and_decode(program, package, prover_name)?;
+
+        match debug_result {
+            Ok((return_value, witness_stack)) => {
+                let witness_stack_result = witness_stack.clone();
+                println!("[{}] Circuit witness successfully solved", package.name);
+
+                if let Some(return_value) = return_value {
+                    println!("[{}] Circuit output: {return_value:?}", package.name);
+                }
+
+                if let Some(witness_name) = witness_name {
+                    let witness_path =
+                        match save_witness_to_dir(witness_stack, witness_name, target_dir) {
+                            Ok(path) => path,
+                            Err(err) => return Err(CliError::from(err)),
+                        };
+
+                    println!("[{}] Witness saved to {}", package.name, witness_path.display());
+                }
+                Ok(Ok(witness_stack_result))
             }
+            Err(error) => Ok(Err(error)),
+        }
+    })
+}
 
-            if let Some(witness_name) = witness_name {
-                let witness_path =
-                    save_witness_to_dir(solved_witness_stack, witness_name, target_dir)?;
-
-                println!("[{}] Witness saved to {}", package.name, witness_path.display());
-            }
-        } else {
-            println!("Debugger execution halted.");
+fn get_test_function(
```

Comment:
This function (or most of it) can be moved into `nargo::ops` and reused in `test_cmd`.